pkg/nats/stream/producer_client.go

```diff
@@ -0,0 +1,204 @@
+		delete(pc.activeStreamInfo, connDetails.ConnID)
+		delete(pc.activeConnHeartBeatRequestSubjects, connDetails.ConnID)
+	}
+
+	if len(pc.activeStreamInfo) == 0 && pc.heartBeatCancelFunc != nil {
+		pc.heartBeatCancelFunc()
+		pc.heartBeatCancelFunc = nil
+	}
+}
+
+func (pc *ProducerClient) heartBeat(ctx context.Context) {
+	ctxWithCancel, cancel := context.WithCancel(ctx)
+	pc.heartBeatCancelFunc = cancel
+
+	ticker := time.NewTicker(pc.config.HeartBeatConfig.HeartBeatIntervalDuration)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctxWithCancel.Done():
+			log.Ctx(ctxWithCancel).Debug().Msgf("Heart beat for producer client %s", pc.ID)
+			return
+		case <-ticker.C:
+
+			nonActiveStreamIds := make(map[string][]string)
+
+			for c, v := range pc.activeConnHeartBeatRequestSubjects {
+				// Create an empty slice for activeStreamIds
+				var activeStreamIds []string
+
+				if streamInfoMap, ok := pc.activeStreamInfo[c]; ok {
+					for _, streamInfo := range streamInfoMap {
+						activeStreamIds = append(activeStreamIds, streamInfo.ID)
+					}
+				}
+
+				heartBeatRequest := HeartBeatRequest{
+					ProducerConnID:  pc.ID,
+					ActiveStreamIds: activeStreamIds,
+				}
+
+				data, err := json.Marshal(heartBeatRequest)
+				if err != nil {
+					log.Err(err)
+					continue
+				}
+
+				msg, err := pc.Conn.Request(v, data, pc.config.HeartBeatConfig.HeartBeatRequestTimeout)
+				if err != nil {
+					nonActiveStreamIds[c] = []string{}
+					continue
+				}
+
+				var heartBeatResponse ConsumerHeartBeatResponse
+				err = json.Unmarshal(msg.Data, &heartBeatResponse)
+				if err != nil {
+					continue
+				}
+				nonActiveStreamIds[c] = heartBeatResponse.NonActiveStreamIds
+			}
+
+			pc.updateActiveStreamInfo(nonActiveStreamIds)
+		}
+	}
+}
+
+func (pc *ProducerClient) WriteResponse(conn *ConnectionDetails, obj interface{}, writer *Writer) (int, error) {
+	pc.mu.Lock()
+	streamIds, active := pc.activeStreamInfo[conn.ConnID]
+
+	for _, v := range streamIds {
+		if v.ID == conn.StreamID {
+			active = true
+		}
+	}
+	pc.mu.Unlock()
+
+	if !active {
+		return 0, fmt.Errorf("stream id is now closed")
+	}
+
+	return writer.WriteObject(obj)
+}
+
+func (pc *ProducerClient) updateActiveStreamInfo(nonActiveStreamIds map[string][]string) {
+	updates := make(map[string]map[string]StreamInfo)
+
+	for connID, nonActiveIdsList := range nonActiveStreamIds {
+		nonActiveMap := make(map[string]bool)
+		for _, id := range nonActiveIdsList {
+			nonActiveMap[id] = true
+		}
+
+		updatedStreams := make(map[string]StreamInfo)
+		for streamID, streamInfo := range pc.activeStreamInfo[connID] {
+			if _, found := nonActiveMap[streamID]; !found ||
+				time.Since(streamInfo.CreatedAt) <= pc.config.StreamCancellationBufferDuration {
+				updatedStreams[streamID] = streamInfo
+			}
+		}
+
+		if len(updatedStreams) > 0 {
+			updates[connID] = updatedStreams
+		}
+	}
+
+	pc.mu.Lock()
+	for c, s := range updates {
+		pc.activeStreamInfo[c] = s
+	}
```

Comment:
please add comments explaining what it is going on, what edge cases are you handling, and why we are adding a buffer. The whole PR can benefit from more comments as these are not simple changes :). 

Also instead of updating the map to only keep the active streams, it might make things more readable to update the map to delete the inactive streams. This can enable more logic in the future for cleaning up deleted streams, and we should add a debug log line for the streams that are being deleted